### PR TITLE
DaemonSet ObservedGeneration

### PR DESCRIPTION
**Pull request by lukaszo**
Extracting ObserverdGeneration part from #31693. It also implements #7328 for DaemonSets.

cc @kargakis

### CODE REVIEW

File: pkg/apis/extensions/types.go [hunk 1]
```diff
@@ -459,6 +459,10 @@ const (
 
 // DaemonSetStatus represents the current status of a daemon set.
 type DaemonSetStatus struct {
+	// ObservedGeneration is the most recent generation observed by the daemon set controller.
```
Comment (kargakis) [inline pkg/apis/extensions/types.go]: Move this in the bottom of the struct similar to the versioned one.

File: pkg/controller/daemon/daemoncontroller.go [hunk 2]
```diff
@@ -553,6 +553,7 @@ func storeDaemonSetStatus(dsClient unversionedextensions.DaemonSetInterface, ds
 
 	var updateErr, getErr error
 	for i := 0; i < StatusUpdateRetries; i++ {
+		toUpdate.Status.ObservedGeneration = ds.Generation
```
Comment (kargakis) [inline pkg/controller/daemon/daemoncontroller.go]: You also need to compare this above in L544
```
ds.Status.ObservedGeneration >= ds.Generation
```

**Comment by kargakis** [comment]
@bgrant0607 for the api review

cc: @kubernetes/sig-apps-misc

File: pkg/controller/daemon/daemoncontroller.go (hunk 2 above)
Comment (lukaszo) [inline pkg/controller/daemon/daemoncontroller.go]: You mean to run update only when ds.Status.ObservedGeneration is smaller than ds.Generation?
Something like:

```go
if ds.Status.ObservedGeneration >= ds.Generation {
  return nil
}

Comment (kargakis) [inline pkg/controller/daemon/daemoncontroller.go]: Should be part of the rest `if` clause

**Comment by lukaszo** [comment]
I've prepared fix for failing tests here https://github.com/kubernetes/kubernetes/pull/39181

File: pkg/controller/daemon/daemoncontroller.go (hunk 2 above)
Comment (lukaszo) [inline pkg/controller/daemon/daemoncontroller.go]: done

File: pkg/controller/daemon/daemoncontroller_test.go [hunk 3]
```diff
@@ -609,3 +609,29 @@ func TestNumberReadyStatus(t *testing.T) {
...
+		}
+		return false, nil, nil
+	})
+
+	addNodes(manager.nodeStore.Store, 0, 1, simpleNodeLabel)
+	addPods(manager.podStore.Indexer, "node-0", simpleDaemonSetLabel, 1)
+	manager.dsStore.Add(daemon)
+
+	syncAndValidateDaemonSets(t, manager, daemon, podControl, 0, 0)
+	fmt.Println(updated.Status)
```
Comment (kargakis) [inline pkg/controller/daemon/daemoncontroller_test.go]: remove this println

**Comment by kargakis** [comment]
LGTM, waiting on an api review from @kubernetes/api-approvers

File: pkg/controller/daemon/daemoncontroller_test.go (hunk 3 above)
Comment (lukaszo) [inline pkg/controller/daemon/daemoncontroller_test.go]: Removed

**Comment by soltysh** [comment]
LGTM as well

**Comment by smarterclayton** [comment]
API changes LGTM, we've stated before this was fundamental to controllers.  Approved.

File: pkg/apis/extensions/v1beta1/types.go [hunk 4]
```diff
@@ -568,6 +568,10 @@ type DaemonSetStatus struct {
 	// NumberReady is the number of nodes that should be running the daemon pod and have one
 	// or more of the daemon pod running and ready.
 	NumberReady int32 `json:"numberReady" protobuf:"varint,4,opt,name=numberReady"`
+
+	// ObservedGeneration is the most recent generation observed by the daemon set controller.
```
Comment (bgrant0607) [inline pkg/apis/extensions/v1beta1/types.go]: All of these comments should reference the json field names rather than the Go field names.

**Pull request merged**